added support for retrieving mementos

pwinckles opened the pull request:
**JIRA Ticket**: [FCREPO-3245](https://jira.lyrasis.org/projects/FCREPO/issues/FCREPO-3245)

# What does this Pull Request do?

Adds the ability to retrieve mementos.

# How should this be tested?

Test script:

```
# Create an AG
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/version-test-3 -H'Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"' -H'Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"' -v -H "Content-Type: text/turtle" -X PUT --data "<> <http://purl.org/dc/elements/1.1/title> 'AG Parent'"

# Create a child of the AG
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/version-test-3/child1 -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "testing"

# Attempt to find a memento
curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -H "Accept-Datetime: Mon, 09 Mar 2020 17:00:00 GMT" -v http://localhost:8080/rest/version-test-3

< HTTP/1.1 406 Not Acceptable
< Set-Cookie: JSESSIONID=node01k7b201ep5d6r153a2qwcopxcy2.node0; Path=/
< Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Mon, 09-Mar-2020 14:56:57 GMT
< Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
< Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"
< Link: <http://www.w3.org/ns/ldp#Container>;rel="type"
< Link: <http://www.w3.org/ns/ldp#RDFSource>; rel="type"
< Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"
< Link: <http://localhost:8080/rest/version-test-3>; rel="timegate"
< Link: <http://localhost:8080/rest/version-test-3>; rel="original"
< Link: <http://localhost:8080/rest/version-test-3/fcr:versions>; rel="timemap"
< Link: <http://mementoweb.org/ns#OriginalResource>; rel="type"
< Link: <http://mementoweb.org/ns#TimeGate>; rel="type"
< Accept-External-Content-Handling: copy,redirect,proxy
< Accept-Patch: application/sparql-update
< Accept-Post: text/turtle,text/rdf+n3,text/n3,application/rdf+xml,application/n-triples,application/ld+json
< Allow: MOVE,COPY,DELETE,POST,HEAD,GET,PUT,PATCH,OPTIONS
< Link: <http://localhost:8080/rest/version-test-3/fcr:acl>; rel="acl"
< Preference-Applied: return=representation
< Cache-Control: must-revalidate,no-cache,no-store
< Content-Length: 0
< Server: Jetty(9.4.24.v20191120)

# Create a memento
curl -u fedoraAdmin:fedoraAdmin -X POST -v http://localhost:8080/rest/version-test-3/fcr:versions

# Find a memento
curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -H "Accept-Datetime: Mon, 09 Mar 2020 17:00:00 GMT" -v http://localhost:8080/rest/version-test-3

< HTTP/1.1 302 Found
< Date: Tue, 10 Mar 2020 14:59:48 GMT
< Set-Cookie: JSESSIONID=node07umaoh56bopz184qx3zggro8f5.node0; Path=/
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Mon, 09-Mar-2020 14:59:48 GMT
< Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
< Link: <http://fedora.info/definitions/v4/repository#ArchivalGroup>;rel="type"
< Link: <http://www.w3.org/ns/ldp#Container>;rel="type"
< Link: <http://www.w3.org/ns/ldp#RDFSource>; rel="type"
< Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"
< Link: <http://localhost:8080/rest/version-test-3>; rel="timegate"
< Link: <http://localhost:8080/rest/version-test-3>; rel="original"
< Link: <http://localhost:8080/rest/version-test-3/fcr:versions>; rel="timemap"
< Link: <http://mementoweb.org/ns#OriginalResource>; rel="type"
< Link: <http://mementoweb.org/ns#TimeGate>; rel="type"
< Accept-External-Content-Handling: copy,redirect,proxy
< Accept-Patch: application/sparql-update
< Accept-Post: text/turtle,text/rdf+n3,text/n3,application/rdf+xml,application/n-triples,application/ld+json
< Allow: MOVE,COPY,DELETE,POST,HEAD,GET,PUT,PATCH,OPTIONS
< Link: <http://localhost:8080/rest/version-test-3/fcr:acl>; rel="acl"
< Preference-Applied: return=representation
< Vary: Prefer
< Vary: Accept
< Vary: Range
< Vary: Accept-Encoding
< Vary: Accept-Language
< Vary: Accept-Datetime
< Location: http://localhost:8080/rest/version-test-3/fcr:versions/20200310145927
< Content-Length: 0
< Server: Jetty(9.4.24.v20191120)

# Retrieve AG memento
curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -v http://localhost:8080/rest/version-test-3/fcr:versions/20200310145927

<info:fedora/version-test-3>
        dc:title             "AG Parent" ;
        rdf:type             ldp:RDFSource ;
        fedora:created       "2020-03-10T14:56:24.560756Z" ;
        fedora:lastModified  "2020-03-10T14:56:24.560756Z" ;
        rdf:type             ldp:BasicContainer ;
        rdf:type             fedora:ArchivalGroup .
		
# Retrieve child memento
curl -u fedoraAdmin:fedoraAdmin -v http://localhost:8080/rest/version-test-3/child1/fcr:versions/20200310145927

< HTTP/1.1 200 OK
< Date: Tue, 10 Mar 2020 15:24:49 GMT
< Set-Cookie: JSESSIONID=node0zapulvibku6wfrke08ex49wv1.node0; Path=/
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Mon, 09-Mar-2020 15:24:49 GMT
< ETag: "7494DAE3016C368D7773144D0CCC929F"
< X-State-Token: 7494DAE3016C368D7773144D0CCC929F
< Last-Modified: Tue, 10 Mar 2020 14:56:43 GMT
< Content-Type: text/plain
< Accept-Ranges: bytes
< Content-Disposition: attachment; filename=""; creation-date="Tue, 10 Mar 2020 14:56:43 GMT"; modification-date="Tue, 10 Mar 2020 14:56:43 GMT"; size=7
< Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
< Link: <http://www.w3.org/ns/ldp#NonRDFSource>;rel="type"
< Link: <http://localhost:8080/rest/version-test-3/child1/fcr:metadata>; rel="describedby"
< Link: <http://localhost:8080/rest/version-test-3/child1>; rel="timegate"
< Link: <http://localhost:8080/rest/version-test-3/child1>; rel="original"
< Link: <http://localhost:8080/rest/version-test-3/child1/fcr:versions>; rel="timemap"
< Accept-External-Content-Handling: copy,redirect,proxy
< Allow: GET,HEAD,OPTIONS,DELETE
< Memento-Datetime: Tue, 10 Mar 2020 14:59:27 GMT
< Link: <http://mementoweb.org/ns#Memento>; rel="type"
< Cache-Control: no-transform, must-revalidate, max-age=0
< Content-Length: 7
< Server: Jetty(9.4.24.v20191120)
< 
* Connection #0 to host localhost left intact
testing

# Update the child
curl -u fedoraAdmin:fedoraAdmin http://localhost:8080/rest/version-test-3/child1 -H "Link: <http://www.w3.org/ns/ldp#NonRDFSource>; rel=type" -v -H "Content-Type: text/plain" -X PUT --data-binary "updated"

# Create a new memento
curl -u fedoraAdmin:fedoraAdmin -X POST -v http://localhost:8080/rest/version-test-3/fcr:versions

# Retrieve the original memento
curl -u fedoraAdmin:fedoraAdmin -v http://localhost:8080/rest/version-test-3/child1/fcr:versions/20200310145927

testing

# Retrieve the new memento
curl -u fedoraAdmin:fedoraAdmin -v http://localhost:8080/rest/version-test-3/child1/fcr:versions/20200310153910

updated
```

Same outputs from F5:

```
curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -H "Accept-Datetime: Mon, 09 Mar 2020 17:00:00 GMT" -v http://localhost:8080/rest/version-test-3

< HTTP/1.1 406 Not Acceptable
< Date: Tue, 10 Mar 2020 15:28:44 GMT
< Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
< Link: <http://www.w3.org/ns/ldp#Container>;rel="type"
< Link: <http://www.w3.org/ns/ldp#RDFSource>; rel="type"
< Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"
< Link: <http://localhost:8080/rest/version-test-3>; rel="timegate"
< Link: <http://localhost:8080/rest/version-test-3>; rel="original"
< Link: <http://localhost:8080/rest/version-test-3/fcr:versions>; rel="timemap"
< Link: <http://mementoweb.org/ns#OriginalResource>; rel="type"
< Link: <http://mementoweb.org/ns#TimeGate>; rel="type"
< Accept-External-Content-Handling: copy,redirect,proxy
< Accept-Patch: application/sparql-update
< Accept-Post: text/turtle,text/rdf+n3,text/n3,application/rdf+xml,application/n-triples,application/ld+json
< Allow: MOVE,COPY,DELETE,POST,HEAD,GET,PUT,PATCH,OPTIONS
< Link: <http://localhost:8080/rest/version-test-3/fcr:acl>; rel="acl"
< Preference-Applied: return=representation
< Vary: Prefer
< Vary: Accept
< Vary: Range
< Vary: Accept-Encoding
< Vary: Accept-Language
< Vary: Accept-Datetime
< Cache-Control: must-revalidate,no-cache,no-store
< Content-Length: 0
< Server: Jetty(9.3.25.v20180904)


curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -H "Accept-Datetime: Mon, 09 Mar 2020 17:00:00 GMT" -v http://localhost:8080/rest/version-test-3

< HTTP/1.1 302 Found
< Date: Tue, 10 Mar 2020 15:30:14 GMT
< Link: <http://www.w3.org/ns/ldp#Resource>;rel="type"
< Link: <http://www.w3.org/ns/ldp#Container>;rel="type"
< Link: <http://www.w3.org/ns/ldp#RDFSource>; rel="type"
< Link: <http://www.w3.org/ns/ldp#BasicContainer>;rel="type"
< Link: <http://localhost:8080/rest/version-test-3>; rel="timegate"
< Link: <http://localhost:8080/rest/version-test-3>; rel="original"
< Link: <http://localhost:8080/rest/version-test-3/fcr:versions>; rel="timemap"
< Link: <http://mementoweb.org/ns#OriginalResource>; rel="type"
< Link: <http://mementoweb.org/ns#TimeGate>; rel="type"
< Accept-External-Content-Handling: copy,redirect,proxy
< Accept-Patch: application/sparql-update
< Accept-Post: text/turtle,text/rdf+n3,text/n3,application/rdf+xml,application/n-triples,application/ld+json
< Allow: MOVE,COPY,DELETE,POST,HEAD,GET,PUT,PATCH,OPTIONS
< Link: <http://localhost:8080/rest/version-test-3/fcr:acl>; rel="acl"
< Preference-Applied: return=representation
< Vary: Prefer
< Vary: Accept
< Vary: Range
< Vary: Accept-Encoding
< Vary: Accept-Language
< Vary: Accept-Datetime
< Location: http://localhost:8080/rest/version-test-3/fcr:versions/20200310152937
< Content-Length: 0
< Server: Jetty(9.3.25.v20180904)


curl -u fedoraAdmin:fedoraAdmin -H "Accept: text/turtle" -v http://localhost:8080/rest/version-test-3/fcr:versions/20200310152937

<http://localhost:8080/rest/version-test-3>
        rdf:type               fedora:Container ;
        rdf:type               ldp:RDFSource ;
        rdf:type               ldp:Container ;
        rdf:type               fedora:Resource ;
        rdf:type               ldp:BasicContainer ;
        fedora:created         "2020-03-10T15:27:24.684Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> ;
        ldp:contains           <http://localhost:8080/rest/version-test-3/child1> ;
        dc:title               "AG Parent" ;
        fedora:lastModified    "2020-03-10T15:27:40.661Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> ;
        fedora:lastModifiedBy  "bypassAdmin" ;
        fedora:createdBy       "bypassAdmin" .
```

# Interested parties
@awoods

awoods commented:
Thanks, @pwinckles !
I look forward to giving this a review.

pwinckles commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java:
```diff
@@ -249,6 +247,26 @@ public void getTimeMapFromAgWithChildrenWithDifferentVersions() throws Exception
+
+        final HttpGet httpGet = new HttpGet(subjectUri + "/child1/fcr:versions" + mementoTime);
+        try (final CloseableHttpResponse response = execute(httpGet)) {
+            assertMementoDatetimeHeaderMatches(response, now());
```
The assertion logic just makes sure that the memento timestamp is within a few seconds of the specified time because it's not possible to specify a timestamp anymore.

pwinckles commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraBaseResource.java:
```diff
@@ -59,6 +63,8 @@
 
     private static final Pattern TRAILING_SLASH_REGEX = Pattern.compile("/+$");
 
+    private static final Pattern MEMENTO_PATH_PATTERN = Pattern.compile(".*/fcr:versions/(.*)$");
```
Yes, to using the constant, but no to `\\d`. We actually don't want the more specific version here, because we have logic for returning helpful error messages for malformed mementos. I want it to match anything with `fcr:versions` regardless of if it's correct, so that I can throw an error if I can't interpret it as a memento. I'll leave a comment so that it's clear this is intended, though integration tests would break if it was changed.

pwinckles commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java:
```diff
@@ -989,52 +926,35 @@ public void testCreateVersionOfBinary() throws Exception {
-        final HttpGet httpGet = new HttpGet(mementoUri);
-        try (final CloseableHttpResponse response = execute(httpGet)) {
-            assertMementoDatetimeHeaderPresent(response);
+            // Content-type is not retained for a binary memento created without description
```
I'm not sure what that means... It only looks like I added it because of other editing I did in the area.

https://github.com/fcrepo4/fcrepo4/blob/f02d2e637e6364db133bddbe759470e7a7c7d4d1/fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java#L1039

??

awoods replied:
I suppose we should ensure that the content-type for binary mementos is retained... then the comment can go away :)

pwinckles commented on fcrepo-kernel-impl/src/test/java/org/fcrepo/kernel/impl/models/FedoraResourceImplTest.java:
```diff
@@ -0,0 +1,127 @@
+        final var resource = resourceWithMockedTimeMap(ID);
+        expectMementos("20200309172117", "20200309172117", "20200309172117");
+        final var match = resource.findMementoByDatetime(instant("20200309172118"));
+        assertEquals("2", match.getId());
```
There isn't currently a way to do so.

awoods replied:
If we can not subsequently retrieve a memento, maybe we should throw an exception during its creation.

pwinckles commented on fcrepo-http-api/src/test/java/org/fcrepo/integration/http/api/FedoraVersioningIT.java:
```diff
@@ -1206,21 +1127,25 @@ public void testGetVersionResponseContentTypes() throws Exception {
+
+        TimeUnit.SECONDS.sleep(1);
+
+        final String version2Uri = createLDPRSMementoWithExistingBody();
```
Hmm... Yes, I'm confused now too. I'll see what's going on here tomorrow.

pwinckles replied:
Ah. So, what's going on here is that the most recent version of the resource is already tagged with a memento (ie there is no mutable head); so a new memento is not created. Instead, the service returns the location of the most recent memento.

pwinckles replied:
I'll fix the test so that it's testing what it's supposed to be testing, but it's an open question if you're alright with the behavior I just described.

pwinckles commented:
I updated the PR addressing your comments. Open questions:

1. What should happen if a create memento operation results in a multiple mementos with the same datetime?
1. What should happen if a create memento operation is executed on a resource that has had no changes since the most recent memento was created?

My answers to those questions are:

1. Allow the creation of the mementos, and always return the version with the highest OCFL version number when the memento datetime is retrieved. I suspect that this kind of collision will be very rare outside of test cases, and I can't think of a good way to prevent it.
1. Do no create a new version. Return success with a reference to the existing memento, which is the head state of the resource.

pwinckles commented:
Furthermore, let's consider the behavior of F5. If I update a resource multiple times within the same second, and then create a memento on that second, the only state of the resource that I'll be able to access is the most recent state within the second. Or at least I assume that's the case. The only difference here is that OCFL can offer a higher level of granularity (though it would not preform well with so many versions), but it's just not exposed through the API.

awoods commented:
> 1. Allow the creation of the mementos, and always return the version with the highest OCFL version number when the memento datetime is retrieved. I suspect that this kind of collision will be very rare outside of test cases, and I can't think of a good way to prevent it.
> 2. Do no create a new version. Return success with a reference to the existing memento, which is the head state of the resource.

I agree with the answer/approach for "2.".

For "1.", is there a way of rejecting a request to commit a memento if there is already a memento within the same second? If so, that seems more clear (least surprise) from the user's perspective.

pwinckles commented:
The memento timestamp is the time the OCFL version is created, which is created by the OCFL client. The only way that I can think of to prevent multiple versions from being created in the same second requires using the OCFL client's optimistic locking feature. It would look something like this.

1. Request received to create a memento
1. The corresponding OCFL object is loaded into a Fedora OCFL object session and the current HEAD OCFL version of the object and its created timestamp are noted
1. When the object session is committed:
   1. If `now` is within the same second as the recorded timestamp, then the request is rejected (alternatively, sleep for a second)
   1. If `now` is not within the same second, then the new version is created referencing the HEAD version it recorded when the session was opened. The OCFL client will fail the request if the referenced version is no longer the HEAD version (ie another version was created between the time the object was first loaded and the time the session was committed). If an intermediary version was created but there is not a timestamp collision, version creation will still fail. The only way around that problem is if Fedora's OCFL session wanted to implement retry logic.

Currently, optimistic locking is not used, and Fedora will always attempt to update an OCFL object regardless of its state when the OCFL object session was opened.

Personally, I don't think it's worth it, and I don't think the current behavior is particularly surprising. The memento API has second level granularity. If you attempt to create multiple mementos within the same second, then only the last memento created will be accessible.

awoods commented:
@pwinckles : Thanks for the detail. OK... let's proceed as implemented. 

I have updated the documentation ticket to reflect the need to mention the multiple-versions-within-the-same-second behavior.
https://jira.lyrasis.org/browse/FCREPO-3242

pwinckles commented:
This review is listed as changes request. I think I addressed all of your concerns, did I miss one?

whikloj commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java:
```diff
@@ -291,7 +291,7 @@ private RdfStream getResourceTriples(final int limit, final FedoraResource resou
             //streams.add(getTriples(resource, MINIMAL));
             // Mementos already have the server managed properties in the PROPERTIES category
             // since mementos are immutable and these triples are no longer managed
```
Note can probably go if the check for Mementos is.

pwinckles replied:
Done

whikloj commented on fcrepo-http-api/src/main/java/org/fcrepo/http/api/ContentExposingResource.java:
```diff
@@ -302,7 +302,7 @@ private RdfStream getResourceTriples(final int limit, final FedoraResource resou
             // Additional server-managed triples about this resource
             // Mementos already have the server managed properties in the PROPERTIES category
             // since mementos are immutable and these triples are no longer managed
```
Note can go.

pwinckles replied:
Done
